Update op-geth dependency to upstream pre-v1.15.0 master merge & ignore SetCode batcher txs

sebastianst opened the pull request:
**Description**

This is the corresponding monorepo PR to the op-geth upstream merge happening at https://github.com/ethereum-optimism/op-geth/pull/474.

Crucially, this update gives us access to the new EIP-7702 `SetCode` transaction type.

The PR has to adapt a bunch of API usages due to upstream changes. Reviewers please be aware of the TODO notes.

One soft-fork consensus rule is added: only Legacy, ACL, Dynamic, Blob, Deposit transactions are considered valid batcher transaction candidates. SetCode and any future transaction types are skipped (for now).

geth also doesn't resolve DNS names in legacy enode URLs any more, so this IP resolution has been added to the op-node's p2p init startup code.

**Metadata**

Towards https://github.com/ethereum-optimism/optimism/issues/13627

sebastianst committed:
go: Update op-geth dependency to upstream pre-v1.14.13-master merge

We need the 7702 SetCode transaction type.

sebastianst committed:
op-node/rollup/derive: treat SetCode or future tx types as invalid batcher txs

sebastianst committed:
all: adapt to op-geth upstream changes

tynes commented on op-node/rollup/derive/data_source.go:
```diff
@@ -91,9 +91,15 @@ type DataSourceConfig struct {
+//  2. the transaction has a To() address that matches the batch inbox address, and
+//  3. the transaction has a valid signature from the batcher address
 func isValidBatchTx(tx *types.Transaction, l1Signer types.Signer, batchInboxAddr, batcherAddr common.Address, logger log.Logger) bool {
+	// For now, we want to disallow the SetCodeTx type or any future types.
```
We should revisit if this is necessary still. This rule was introduced to save the work of needing to know about the serialization of the SetCodeTx. Since that approach didn't work, not sure if we need to add this extra rule given we did the work for the rebase. My reasoning is to reduce branching logic that needs additional test coverage. An honest batcher just won't submit other tx types

sebastianst replied:
OTOH this rule prepares forwards-compatibility, as we have specified it in [this specs update](https://github.com/ethereum-optimism/specs/pull/518). So having this rule in now means we can handle generic future tx type and receipt type handling as a soft fork in the future. I don't see a use case for `SetCode` batcher txs  right now.

Without this rule, I think adding future-proof transaction type handling is harder.

sebastianst replied:
[Decision was made](https://github.com/ethereum-optimism/pm/issues/28) to keep this rule in, but also allow Deposit transactions. Updated.

sebastianst replied:
I added a test to the blob source tests that a SetCode tx is processed and dropped, that otherwise looks like a legit batcher tx.

sebastianst committed:
go: Update op-geth dependency again to include latest changes

sebastianst committed:
derive: allow Deposit batcher txs

This allows batch tx force-inclusion for L3s.

sebastianst commented on op-chain-ops/foundry/allocs_test.go:
```diff
@@ -58,15 +58,15 @@ func TestForgeAllocs_FromState(t *testing.T) {
 	// Commit and make a new state, we cannot reuse the state after Commit
 	// (see doc-comment in Commit, absolute footgun)
-	root, err := st.Commit(0, false)
+	root, err := st.Commit(0, false, false) // TODO: false ok here? In general, should be IsCancun()
```
TBD

protolambda replied:
It's `false` in the geth `genesis.go`, so let's go with that. Pruning empty accounts from genesis allocs makes sense to me.

sebastianst commented on op-chain-ops/foundry/allocs_test.go:
```diff
@@ -58,15 +58,15 @@ func TestForgeAllocs_FromState(t *testing.T) {
 	st.SetState(contract, crypto.Keccak256Hash([]byte("hello")), crypto.Keccak256Hash([]byte("world")))
 
-	root, err = st.Commit(0, false)
+	root, err = st.Commit(0, false, false) // TODO: false ok here? In general, should be IsCancun()
```
TBD

protolambda replied:
same thing as above

sebastianst commented on op-chain-ops/cmd/op-simulate/main.go:
```diff
@@ -295,7 +294,10 @@ func simulate(ctx context.Context, logger log.Logger, conf *params.ChainConfig,
 	start := time.Now()
-	receipt, err := core.ApplyTransaction(conf, cCtx, &sender, &gp, state, header, tx, &usedGas, vmConfig)
+	// TODO: is a nil author ok?
+	blockCtx := core.NewEVMBlockContext(header, cCtx, nil, conf, state)
+	evm := vm.NewEVM(blockCtx, state, conf, vmConfig)
+	receipt, err := core.ApplyTransaction(evm, &gp, state, header, tx, &usedGas)
```
TBD

sebastianst commented on op-chain-ops/script/script.go:
```diff
@@ -697,7 +687,9 @@ func (h *Host) StateDump() (*foundry.ForgeAllocs, error) {
-	root, err := baseState.Commit(h.env.Context.BlockNumber.Uint64(), true)
+	// TODO: do we need to check `IsCancun()` here, or can we just pass false?
+	isCancun := h.chainCfg.IsCancun(h.env.Context.BlockNumber, h.env.Context.Time)
+	root, err := baseState.Commit(h.env.Context.BlockNumber.Uint64(), true, isCancun)
```
TBD

protolambda replied:
Just using `false` here is fine, we assume cancun is active everywhere here, and there's no historical sync in the script host environment.

sebastianst commented on op-e2e/actions/helpers/l1_miner.go:
```diff
@@ -177,8 +177,11 @@ func (s *L1Miner) IncludeTx(t Testing, tx *types.Transaction) {
 	s.l1BuildingState.SetTxContext(tx.Hash(), len(s.L1Transactions))
-	receipt, err := core.ApplyTransaction(s.l1Cfg.Config, s.l1Chain, &s.l1BuildingHeader.Coinbase,
-		s.L1GasPool, s.l1BuildingState, s.l1BuildingHeader, tx.WithoutBlobTxSidecar(), &s.l1BuildingHeader.GasUsed, *s.l1Chain.GetVMConfig())
+	blockCtx := core.NewEVMBlockContext(s.l1BuildingHeader, s.l1Chain, nil, s.l1Cfg.Config, s.l1BuildingState)
+	// TODO: we could reuse the evm for the same block and use SetTxContext here
+	evm := vm.NewEVM(blockCtx, s.l1BuildingState, s.l1Cfg.Config, *s.l1Chain.GetVMConfig())
```
TBD

protolambda replied:
I think it's simpler to not reuse the EVM across txs here. And ApplyTransaction runs the tx-context setter (see other comment).

sebastianst commented on op-program/client/l2/db_test.go:
```diff
@@ -143,7 +143,9 @@ func TestUpdateState(t *testing.T) {
 	require.Equal(t, []byte{1}, statedb.GetCode(codeAccount))
 
 	// Changes should be available under the new state root after committing
-	newRoot, err := statedb.Commit(genesisBlock.NumberU64()+1, false)
+	isCancun := l2Genesis.Config.IsCancun(genesisBlock.Number(), genesisBlock.Time())
+	// TODO: can we just use false instead of checking IsCancun()?
+	newRoot, err := statedb.Commit(genesisBlock.NumberU64()+1, false, isCancun)
```
TBD

protolambda replied:
It's a test, so I think yes. But I am going to prefer `isCancun` check still, since it'll be consistent with other op-program code, and should be safe that way

sebastianst commented on op-program/client/l2/engineapi/block_processor.go:
```diff
@@ -128,8 +128,16 @@ func (b *BlockProcessor) CheckTxWithinGasLimit(tx *types.Transaction) error {
 	b.state.SetTxContext(tx.Hash(), txIndex)
-	receipt, err := core.ApplyTransaction(b.dataProvider.Config(), b.dataProvider, &b.header.Coinbase,
-		b.gasPool, b.state, b.header, tx, &b.header.GasUsed, *b.dataProvider.GetVMConfig())
+
+	context := core.NewEVMBlockContext(b.header, b.dataProvider, nil, b.dataProvider.Config(), b.state)
+	// TODO: Does this need to happen?
+	var precompileOverrides vm.PrecompileOverrides
+	if vmConfig := b.dataProvider.GetVMConfig(); vmConfig != nil && vmConfig.PrecompileOverrides != nil {
+		precompileOverrides = vmConfig.PrecompileOverrides
+	}
```
TBD - applying the overrides was taken from the BlockProcessor ctor `NewBlockProcessorFromHeader`, where the evm is constructed in this way.

sebastianst replied:
why don't we just pass `b.dataProvider.GetVMConfig()` to `vm.NewEVM`?

protolambda replied:
We should do that. I checked the usages, and it's never nil. And if it is nil, we'll get a very loud warning on the first test that runs the EVM there.

sebastianst commented on op-program/client/l2/engineapi/block_processor.go:
```diff
@@ -128,8 +128,16 @@ func (b *BlockProcessor) CheckTxWithinGasLimit(tx *types.Transaction) error {
+	if vmConfig := b.dataProvider.GetVMConfig(); vmConfig != nil && vmConfig.PrecompileOverrides != nil {
+		precompileOverrides = vmConfig.PrecompileOverrides
+	}
+	// TODO: reuse evm?
+	evm := vm.NewEVM(context, b.state, b.dataProvider.Config(), vm.Config{PrecompileOverrides: precompileOverrides})
```
Note that with the upstream refactor, it is suggested to reuse the EVM instance and update it for new tx application with `SetTxContext`. But this should still work afaiu. We could move the reuse of the EVM instance here and in the action tests to a follow-up.

protolambda replied:
Opened an issue to track the EVM reuse: https://github.com/ethereum-optimism/optimism/issues/14038

I think it's a nice change, but I don't want to break any op-program assumptions with optimizations, or delay the geth merge with op-program work, even if simple.

sebastianst commented on op-wheel/cheat/cheat.go:
```diff
@@ -324,7 +325,8 @@ func StoragePatch(patch io.Reader, address common.Address) HeadFn {
 			i += 1
 			if i%1000 == 0 { // for every 1000 values, commit to disk
-				if _, err := headState.Commit(head.Number.Uint64(), true); err != nil {
+				// TODO: ok to just pass `false` here?
+				if _, err := headState.Commit(head.Number.Uint64(), true, false); err != nil {
```
TBD

protolambda replied:
yes that's fine here, it's a test tool, ~and all account attributes are written per account at once, so we can assume empty account wiping does no harm.~ edit: not the case, we should not be wiping intermediate storage if the account is unset. Although it would still wipe at the end. Since only storage is changed, and the account attributes may still be empty.

sebastianst reviewed (commented):
Adding notes for reviewers.

sebastianst committed:
Merge remote-tracking branch 'origin/develop' into seb/op-geth-upstream-pre-v1.14.13-master

sebastianst committed:
lint

sebastianst commented on op-chain-ops/script/script.go:
```diff
@@ -261,16 +261,6 @@ func NewHost(
 		Random:      &executionContext.PrevRandao,
 	}
 
-	// Initialize a transaction-context for the EVM to access environment variables.
-	// The transaction context (after embedding inside of the EVM environment) may be mutated later.
-	txContext := vm.TxContext{
-		Origin:       executionContext.Origin,
-		GasPrice:     big.NewInt(0),
-		BlobHashes:   executionContext.BlobHashes,
-		BlobFeeCap:   big.NewInt(0),
-		AccessEvents: state.NewAccessEvents(h.baseState.PointCache()),
-	}
```
OK this probably needs to be set again :D

sebastianst committed:
add txContext back

sebastianst commented on op-chain-ops/cmd/op-simulate/main.go:
```diff
@@ -297,7 +296,10 @@ func simulate(ctx context.Context, logger log.Logger, conf *params.ChainConfig,
 	start := time.Now()
-	receipt, err := core.ApplyTransaction(conf, cCtx, &sender, &gp, state, header, tx, &usedGas, vmConfig)
+	// TODO: is a nil author ok?
+	blockCtx := core.NewEVMBlockContext(header, cCtx, nil, conf, state)
+	evm := vm.NewEVM(blockCtx, state, conf, vmConfig)
+	receipt, err := core.ApplyTransaction(evm, &gp, state, header, tx, &usedGas)
```
Wondering if the `txContext` has to be set here?

protolambda replied:
No. The EVM tx context is set here: `ApplyTransaction` calls `ApplyTransactionExtended` calls `ApplyTransactionWithEVM` calls `ApplyMessage` calls `evm.SetTxContext`

And the state tx context (tx-hash, for logs collection) is already being set

sebastianst commented on op-e2e/actions/helpers/l1_miner.go:
```diff
@@ -177,8 +177,11 @@ func (s *L1Miner) IncludeTx(t Testing, tx *types.Transaction) {
 	s.l1BuildingState.SetTxContext(tx.Hash(), len(s.L1Transactions))
-	receipt, err := core.ApplyTransaction(s.l1Cfg.Config, s.l1Chain, &s.l1BuildingHeader.Coinbase,
-		s.L1GasPool, s.l1BuildingState, s.l1BuildingHeader, tx.WithoutBlobTxSidecar(), &s.l1BuildingHeader.GasUsed, *s.l1Chain.GetVMConfig())
+	blockCtx := core.NewEVMBlockContext(s.l1BuildingHeader, s.l1Chain, nil, s.l1Cfg.Config, s.l1BuildingState)
+	// TODO: we could reuse the evm for the same block and use SetTxContext here
+	evm := vm.NewEVM(blockCtx, s.l1BuildingState, s.l1Cfg.Config, *s.l1Chain.GetVMConfig())
+	receipt, err := core.ApplyTransaction(
+		evm, s.L1GasPool, s.l1BuildingState, s.l1BuildingHeader, tx.WithoutBlobTxSidecar(), &s.l1BuildingHeader.GasUsed)
```
Wondering if the txContext has to be set here too. Old code inside `core.ApplyTransaction` was
```
	msg, err := TransactionToMessage(tx, types.MakeSigner(config, header.Number, header.Time), header.BaseFee)
	if err != nil {
		return nil, err
	}
	txContext := NewEVMTxContext(msg)
```
Note that tests pass though.

protolambda replied:
See other comment, the `ApplyTransaction` still sets tx context.

sebastianst commented on op-program/client/l2/engineapi/block_processor.go:
```diff
@@ -128,8 +128,16 @@ func (b *BlockProcessor) CheckTxWithinGasLimit(tx *types.Transaction) error {
+		precompileOverrides = vmConfig.PrecompileOverrides
+	}
+	// TODO: reuse evm?
+	evm := vm.NewEVM(context, b.state, b.dataProvider.Config(), vm.Config{PrecompileOverrides: precompileOverrides})
+	receipt, err := core.ApplyTransaction(evm, b.gasPool, b.state, b.header, tx, &b.header.GasUsed)
```
Need to `SetTxContext`? Note that tests pass.

protolambda replied:
No need to, see other comments.

sebastianst committed:
fix SetCode batcher tx test

sebastianst commented:
Created tag & image `op-node/v1.10.4-dev.0` at aaf08bfc1d

sebastianst committed:
go: update op-geth dependency to v1.101500.0-dev.1

which is tagged at https://github.com/ethereum-optimism/op-geth/pull/480

sebastianst commented:
Created tag & image `op-node/v1.10.4-dev.1` at 43e17191cd

protolambda committed:
geth: pre-v1.14.13 integration review fixes

protolambda committed:
op-chain-ops: simplify state commit call

sebastianst committed:
op-node/p2p: Use IP address for first Conduit bootnode

After updating to the latest upstream geth dependency, it seems that
host names are no longer supported for bootnodes.

protolambda commented on op-node/p2p/cli/load_config.go:
```diff
@@ -207,6 +217,30 @@ func loadDiscoveryOpts(conf *p2p.Config, ctx *cli.Context) error {
+		return "", fmt.Errorf("looking up IP for hostname %q: %w", u.Hostname(), err)
+	}
+	ip = ips[0]
+	// TODO: add the following check that was in geth for an IPv4 address?
+	// Ensure the IP is 4 bytes long for IPv4 addresses.
+	// if ipv4 := ip.To4(); ipv4 != nil {
+	// 	ip = ipv4
+	// }
```
Yes, I think we should add this. But not only in this part, but also the `net.ParseIP` early return part, to match the original geth behavior. To turn IPs of enodes that do contain an IP also into a v4 IP if possible.

Also, discv5 does support ipv6 I believe, but I think the IPv4-formatted IP is preferable by the enode code.

sebastianst replied:
done

protolambda commented on op-node/p2p/cli/load_config.go:
```diff
@@ -207,6 +217,30 @@ func loadDiscoveryOpts(conf *p2p.Config, ctx *cli.Context) error {
 	return nil
 }
 
+func resolveURLIP(rawurl string) (string, error) {
```
Would be great to add a unit-test with the conduit enode as test-vector.
And then make the `net.LookupIP` function an argument to this helper function, so we can mock the DNS lookup part during testing, to avoid flakes.

sebastianst replied:
added

sebastianst committed:
op-node/p2p: look up DNS host names of enode records

protolambda commented on op-node/p2p/cli/load_config_test.go:
```diff
@@ -0,0 +1,58 @@
+	return nil, errors.New("no such host")
+}
+
+func TestResolveURLIP(t *testing.T) {
```
Very nice

sebastianst committed:
op-node/p2p: Also lookup IPs for default bootnode URLs

sebastianst commented:
Latest tag `op-node/v1.10.4-dev.5` at 339fdfa11d

sebastianst committed:
Merge branch 'develop' into seb/op-geth-upstream-pre-v1.14.13-master

sebastianst committed:
Update op-geth dependency to merged@v1.101500.0-rc.1

sebastianst committed:
op-node/p2p: Add test for default bootnodes
